### DISCUSSION TITLE
Simultaneous connect, close

### DISCUSSION THREAD

**Discussion by junhyeok-yoon** [ANLAB-KAIST/KENSv3#159] [Project Q&A 2. Transport Layer (TCP)] [answered]
In the handshake-simultaneousConnect case,
As far as I know, both sides start connecting at the same time. Consider that one side sends SYN and receives SYN. Then, the connection will be established. And it will be the same to the other side.

1. However, when I see the solution packet log, there are also ACKs which are response to the SYN. Should I ignore it?

2. I cannot infer why second SYN packet of below log has peculiar SEQ#
![vcxsrv_qAGkEyEJWj](https://github.com/ANLAB-KAIST/KENSv3/assets/114990411/360402bc-5cb5-4c20-96c6-de2640f181ef)

3. After established, app can close, so it can send FINACK how should I response with the FINACK? I wonder how should I serve ack# and seq# of ACK packets.

4. What I want to know is HOW to handle the seq# and ack# after the connection established.

4. In handshae-before/after accept, the order of FINACKs, ACKs
Before: FINACK ACK FINACK ACK
After: FINACK FINAXK ACK ACK
Each of the case is normal close and simultaneous close respectively? Also, I cannot understand how ack#, and seq# are manipulated.
(3~5번 질문 모두 seq#와 ack#에 대한 것입니다. 기존의 정상적인 handshake이라면 syn을 보내면서 seq#를 램덤으로 설정하고 synack을 보내면서 상대방의 seq#를 정합니다. 하지만 simultaneous connect을 하면 Syn을 보나면서 seq#를 랜덤하게 설정하는데, 또 그 뒤에 synack을 받기 때문에 이경우에는 최초 설정한 seq#를 유지해야하는건지 아님 어떻게 해야하는지 궁금합니다.)

**Comment by Pusnow** [comment, marked as answer]
For SEQ# and ACK# for simultaneous connection see:

https://github.com/ANLAB-KAIST/KENSv3/blob/eba97b6c81103e365e14e3c13194ccbc2a5966f3/Documentation/rfc/rfc793.txt#L2075-L2093



> 1. However, when I see the solution packet log, there are also ACKs which are response to the SYN. Should I ignore it?

It seems a bug in the reference TCP implementation. You can ignore it.

> 2. I cannot infer why second SYN packet of below log has peculiar SEQ#

The SEQ# is randomly assigned for security. It is a common practice in the real TCP networking stacks.

> 3. After established, app can close, so it can send FINACK how should I response with the FINACK? I wonder how should I serve ack# and seq# of ACK packets.

The FIN packets increase SEQ# by 1.

See:

https://github.com/ANLAB-KAIST/KENSv3/blob/eba97b6c81103e365e14e3c13194ccbc2a5966f3/Documentation/rfc/rfc793.txt#L2484-L2526



> 4. What I want to know is HOW to handle the seq# and ack# after the connection established.

I cannot understand what are you asking. If you meant how to decide SEQ# and ACK#, read:

https://github.com/ANLAB-KAIST/KENSv3/blob/eba97b6c81103e365e14e3c13194ccbc2a5966f3/Documentation/rfc/rfc793.txt#L1597



> 5. In handshae-before/after accept, the order of FINACKs, ACKs
> Before: FINACK ACK FINACK ACK
> After: FINACK FINAXK ACK ACK
> Each of the case is normal close and simultaneous close respectively? Also, I cannot understand > how ack#, and seq# are manipulated.

BeforeAccept is normal close, and AfterAccept is simultaneous close.


> (3~5번 질문 모두 seq#와 ack#에 대한 것입니다. 기존의 정상적인 handshake이라면 syn을 보내면서 seq#를 램덤으로 설정하고 synack을 보내면서 상대방의 seq#를 정합니다. 하지만 simultaneous connect을 하면 Syn을 보나면서 seq#를 랜덤하게 설정하는데, 또 그 뒤에 synack을 받기 때문에 이경우에는 최초 설정한 seq#를 유지해야하는건지 아님 어떻게 해야하는지 궁금합니다.)

Since you have received SYN before sending SYN/ACK, you have already know peer's SEQ#, so you can use it to decide ACK#.

**Reply by junhyeok-yoon** [reply]
Thank you so much!
